visivo/query/insight_tokenizer.py

```diff
@@ -0,0 +1,499 @@
+import re
+from typing import Dict, List, Set, Any, Optional
+import sqlglot
+from sqlglot import exp
+
+from visivo.models.insight import Insight, InsightInteraction
+from visivo.models.sources.source import Source
+from visivo.models.models.model import Model
+from visivo.models.models.local_merge_model import LocalMergeModel
+from visivo.models.tokenized_insight import TokenizedInsight
+from visivo.models.trace_columns import TraceColumns
+from visivo.models.trace_props.layout import Layout
+from visivo.models.trace_props.trace_props import TraceProps
+from visivo.models.base.base_model import BaseModel
+from visivo.models.base.query_string import QueryString
+from visivo.query.dialect import Dialect
+from visivo.query.statement_classifier import StatementClassifier, StatementEnum
+from visivo.utils import extract_value_from_function
+from visivo.logger.logger import Logger
+
+
+class InsightTokenizer:
+    """
+    Tokenizes insights to generate both server-side (pre) and client-side (post) queries.
+    Uses SQLglot for intelligent query analysis and dependency detection.
+    """
+
+    def __init__(self, insight: Insight, model: Model, source: Source):
+        self.insight = insight
+        self.source = source
+        self.model = model
+        self.dialect = Dialect(type=source.type)
+        self.statement_classifier = StatementClassifier(dialect=self.dialect)
+
+        # Analysis results
+        self.select_items = {}  # prop_path -> sql_expression
+        self.column_items = {}  # column_name -> sql_expression
+        self.interaction_dependencies = {}  # interaction -> dependencies
+        self.input_dependencies = set()  # All input names referenced
+        self.required_columns = set()  # Columns needed in pre-query
+        self.groupby_statements = set()  # GROUP BY expressions needed
+
+        # Initialize analysis
+        self._analyze_insight()
+
+    def tokenize(self) -> TokenizedInsight:
+        """Main entry point - returns tokenized insight with pre/post queries"""
+        pre_query = self._generate_pre_query()
```

Comment:
We need to be able to generate these two items very well, I don't think we need to store the tokenized values like we did for the trace tokenizer.